components/cdc/src/delegate.rs

```diff
@@ -45,36 +50,57 @@ impl DownstreamID {
 pub struct Downstream {
     // TODO: include cdc request.
     /// A unique identifier of the Downstream.
-    pub id: DownstreamID,
+    id: DownstreamID,
     // The IP address of downstream.
     peer: String,
     region_epoch: RegionEpoch,
-    sink: UnboundedSender<ChangeDataEvent>,
+    sink: Option<BatchSender<(usize, Event)>>,
 }
 
 impl Downstream {
     /// Create a Downsteam.
     ///
     /// peer is the address of the downstream.
     /// sink sends data to the downstream.
-    pub fn new(
-        peer: String,
-        region_epoch: RegionEpoch,
-        sink: UnboundedSender<ChangeDataEvent>,
-    ) -> Downstream {
+    pub fn new(peer: String, region_epoch: RegionEpoch) -> Downstream {
         Downstream {
             id: DownstreamID::new(),
             peer,
-            sink,
             region_epoch,
+            sink: None,
         }
     }
 
-    fn sink(&self, change_data: ChangeDataEvent) {
-        if self.sink.unbounded_send(change_data).is_err() {
+    pub fn sink_event(&self, change_data_event: Event, size: usize) {
+        if self
+            .sink
+            .as_ref()
+            .unwrap()
+            .send((size, change_data_event))
+            .is_err()
+        {
             error!("send event failed"; "downstream" => %self.peer);
         }
     }
+
+    pub fn set_sink(&mut self, sink: BatchSender<(usize, Event)>) {
+        self.sink = Some(sink);
+    }
+
+    pub fn get_id(&self) -> DownstreamID {
+        self.id
+    }
+
+    pub fn sink_duplicate_error(&self, region_id: u64) {
+        let mut change_data_event = Event::default();
+        let mut cdc_err = EventError::default();
+        let mut err = ErrorDuplicateRequest::default();
+        err.set_region_id(region_id);
+        cdc_err.set_duplicate_request(err);
+        change_data_event.event = Some(Event_oneof_event::Error(cdc_err));
+        change_data_event.region_id = region_id;
+        self.sink_event(change_data_event, 0);
```

Comment:
What does 0 mean? Could you add some comments?

Reply:
The size of error and resolved events is considered zero, I think it's easy to understand so not necessary to add comments for it (It appears in many places).

Reply:
Added the comment of `sink_event`